api/src/main/java/com/epam/pipeline/manager/datastorage/providers/aws/s3/S3Helper.java

```diff
@@ -300,6 +302,54 @@ public void restoreFileVersion(String bucket, String path, String version) {
         moveS3Object(client, bucket, new MoveObjectRequest(path, version, path));
     }
 
+    public void restoreFolder(final String bucket, final String path, final RestoreFolderVO restoreFolderVO) {
+        final AmazonS3 client = getDefaultS3Client();
+        final String requestPath = Optional.ofNullable(path).orElse("");
+        if (!StringUtils.isNullOrEmpty(requestPath)) {
+            Assert.isTrue(checkItemType(client, bucket, requestPath, true) == DataStorageItemType.Folder,
+                    messageHelper.getMessage(MessageConstants.ERROR_FOLDER_INVALID_PATH, requestPath));
+        }
+        try (S3ObjectDeleter deleter = new S3ObjectDeleter(client, bucket)) {
+            restoreFolderFiles(client, bucket, requestPath, restoreFolderVO, deleter);
+        } catch (SdkClientException e) {
+            throw new DataStorageException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private void restoreFolderFiles(final AmazonS3 client, final String bucket, final String path,
+                                    final RestoreFolderVO restoreFolderVO, final S3ObjectDeleter deleter) {
+        getDeletedFiles(client, bucket, ProviderUtils.withTrailingDelimiter(path), restoreFolderVO, deleter)
+                .forEach(item -> deleter.deleteKey(item.getPath(), ((DataStorageFile) item).getVersion()));
+    }
+
+    private List<AbstractDataStorageItem> getDeletedFiles(final AmazonS3 client, final String bucket, final String path,
+                                                          final RestoreFolderVO restoreFolderVO,
+                                                          final S3ObjectDeleter deleter) {
+        return listVersions(client, bucket, path, null, null).getResults().stream()
+                .peek(item -> recursiveRestoreFolderCall(item, client, bucket, restoreFolderVO, deleter))
```

Comment:
I really think that usage of `peek` for some additional operations is not the best option. If we'll change this method to actually delete markers, we can get rid of this call.